project1/server/agent.cpp

```diff
@@ -4,231 +4,415 @@
+        } else if (message.has_terminate()) {
+            return HandleRequest(message.terminate());
+        }
+
+        std::cerr << "No valid message was received." << std::endl;
+        return ClientState::ERROR;
+    }
+
+    bool Agent::SendResponse(const Response &response) {
+        // Serialize the message.
+        if (!wire_protocol::Serialize(response, &outgoing_message_buffer_)) {
+            std::cerr << "Failed to serialize message." << std::endl;
+            return false;
+        }
+
+        // Send the message.
+        if (send(client_fd_, outgoing_message_buffer_.data(),
+                 outgoing_message_buffer_.size(), 0) < 0) {
+            perror("Failed to send the message");
+            return false;
+        }
+        return true;
+    }
+
+    bool Agent::SendFileContents(const ftp_messages::FileContents &file_contents, uint16_t command_id) {
+        // Serialize the message.
+        if (!wire_protocol::Serialize(file_contents, &outgoing_message_buffer_)) {
+            std::cerr << "Failed to serialize message." << std::endl;
+            return false;
+        }
+
+        // monitor for termination request for get and put commands
+        uint16_t totalBytesSent = 0;
+
+        // make 1000 byte buffer
+        int numBytesToSend = 1000;
+        uint8_t buf[numBytesToSend];
+
+        // counter used to track
+        int byte_counter = 0;
+
+        bool terminated = false;
+
+        // continue until we've sent the entire message
+        while (totalBytesSent < outgoing_message_buffer_.size()) {
+
+
+            // fill the buffer one byte at a time
+            for (int n:outgoing_message_buffer_) {
+
+                // check if 1000 byte buffer is filled
+                if (byte_counter == 1000) {
+                    terminated = !active_commands_->Contains(command_id);
+
+                    // check if the process has been terminated
+                    if (!terminated) {
+                        // Send 1000 bytes of the message.
+                        if (send(client_fd_, buf,
+                                 numBytesToSend, 0) < 0) {
+                            perror("Failed to send the file contents.");
+                            return false;
+                        }
+                    } else {
+                        return false;
+                    }
+
+                    // reset
+                    byte_counter = 0;
+
+                    // update
+                    totalBytesSent += 1000;
+                }
+                buf[byte_counter] = outgoing_message_buffer_.at(n);
+                ++byte_counter;
+
+            }
+        }
+        return true;
+    }
+
+    Agent::ClientState Agent::HandleRequest(
+            const ftp_messages::GetRequest &request) {
+        std::cout << "Performing a GET operation." << std::endl;
+
+        Response r;
+
+        // generate a random # for the command id
+        uint32_t id;
```

Comment:
Maybe we should break this ID generation into a separate function?

Reply:
agreed.